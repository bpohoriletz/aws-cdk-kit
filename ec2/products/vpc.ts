import * as ec2 from 'aws-cdk-lib/aws-ec2';

export class VpcProduct implements IVpcProduct {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

interface IVpcProduct extends ec2.VpcProps {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

export interface IVpcBuilder {
  setCidr?(name: string): this;
  setAzs?(): this;
  setNatProvider?(): this;
  setNatSubnets?(): this;
  setSubnetConfiguration?(name: string): this;
  getResult(): ec2.VpcProps;
}

export interface IVpcBuilderConstructor {
  new (): IVpcBuilder;
}
