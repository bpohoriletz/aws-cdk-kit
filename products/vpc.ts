import * as ec2 from 'aws-cdk-lib/aws-ec2';

export class VpcProduct implements IVpcProduct {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

interface IVpcProduct extends ec2.VpcProps {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

export interface IVpcBuilder {
  customizeCidr?(name: string): this;
  customizeAzs?(): this;
  customizeNatProvider?(): this;
  customizeNatSubnets?(): this;
  customizeSubnetConfiguration?(name: string): this;
  getResult(): ec2.VpcProps;
}

export interface IVpcBuilderConstructor {
  new (): IVpcBuilder;
}
