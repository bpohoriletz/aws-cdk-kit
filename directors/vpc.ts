import * as ec2 from "aws-cdk-lib/aws-ec2";
import { Construct } from "constructs";
import { IVpcBuilder, IVpcBuilderConstructor } from "../products/vpc";

export default class VpcDirector {
  private builder: IVpcBuilder;

  constructor(builder: IVpcBuilderConstructor){
    this.builder = new builder();
  }

  constructVpc(scope: Construct, id: string): ec2.Vpc {
    const props =  this.builder
                     .customizeAzs()
                     .customizeCidr(scope.node.path + id)
                     .customizeNatProvider()
                     .customizeNatSubnets()
                     .customizeSubnetConfiguration(scope.node.path + id)
                     .getResult();

    return new ec2.Vpc(scope, id, props)
  }
}
