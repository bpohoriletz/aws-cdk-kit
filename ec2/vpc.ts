import * as cdk from "aws-cdk-lib";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as names from "../naming/resources";
import * as conf from "../../config/ec2/vpc";

export function createMinimalVpc(resourceNamePrefix: string[], stack: cdk.Stack): ec2.Vpc {
  const resourceName = names.ec2VpcName(resourceNamePrefix);
  const vpc = new ec2.Vpc(stack, resourceName, {
    ipAddresses: conf.vpcCidrConfig(resourceName),
    maxAzs: 1,
    subnetConfiguration: conf.vpcSubnetConfig(resourceName),
    natGatewayProvider: ec2.NatProvider.instanceV2({
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T2,
        ec2.InstanceSize.NANO
      ),
    }),
    natGatewaySubnets: {
      subnetType: ec2.SubnetType.PUBLIC, // Make NAT instance reside in a public subnet
    },
  });

  return vpc;
}

export function createVpc(resourceNamePrefix: string[], stack: cdk.Stack): ec2.Vpc {
  const resourceName = names.ec2VpcName(resourceNamePrefix);
  const vpc = new ec2.Vpc(stack, resourceName, {
    ipAddresses: conf.vpcCidrConfig(resourceName),
    maxAzs: 2,             // Use 2 availability zones
    subnetConfiguration: conf.vpcSubnetConfig(resourceName),
    natGatewayProvider: ec2.NatProvider.instanceV2({
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T3,
        ec2.InstanceSize.MICRO
      ),
    }),
    natGatewaySubnets: {
      subnetType: ec2.SubnetType.PUBLIC, // Make NAT instance reside in a public subnet
    },
  });

  return vpc;
}

export function createProdVpc(resourceNamePrefix: string[], stack: cdk.Stack): ec2.Vpc {
  const resourceName = names.ec2VpcName(resourceNamePrefix);
  const vpc = new ec2.Vpc(stack, resourceName, {
    ipAddresses: conf.vpcCidrConfig(resourceName),
    subnetConfiguration: conf.vpcSubnetConfig(resourceName),
  });

  return vpc;
}
