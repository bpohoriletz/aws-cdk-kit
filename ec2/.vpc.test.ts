import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import * as func from "./vpc";

describe("EC2 .createMinimalVpc()", () => {
  let stack: cdk.Stack;

  beforeEach(() => {
    stack = new cdk.Stack();
  });

  test("matches snapshot", async () => {
    func.createMinimalVpc(["pre", "fix"], stack);

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});

describe("EC2 .createVpc()", () => {
  let stack: cdk.Stack;

  beforeEach(() => {
    stack = new cdk.Stack();
  });

  test("matches snapshot", async () => {
    func.createVpc(["pre", "fix"], stack);

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});

describe("EC2 .createProdVpc()", () => {
  let stack: cdk.Stack;

  beforeEach(() => {
    stack = new cdk.Stack();
  });

  test("matches snapshot", async () => {
    func.createProdVpc(["pre", "fix"], stack);

    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
