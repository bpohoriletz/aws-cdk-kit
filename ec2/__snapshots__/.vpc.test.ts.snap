// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EC2 .createMinimalVpc() matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PreFixVpc359F0D17": {
      "Properties": {
        "CidrBlock": "10.255.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "PreFixVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "PreFixVpcIGWB72EC1F8": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PreFixVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "PreFixVpcPublicSubnet1DefaultRouteB716A66A": {
      "DependsOn": [
        "PreFixVpcVPCGWC09F8D3F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet1RouteTable23041149",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PreFixVpcPublicSubnet1RouteTable23041149": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PreFixVpcPublicSubnet1RouteTableAssociation0CC9D25D": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet1RouteTable23041149",
        },
        "SubnetId": {
          "Ref": "PreFixVpcPublicSubnet1Subnet8F335C33",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PreFixVpcPublicSubnet1Subnet8F335C33": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.255.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PreFixVpcVPCGWC09F8D3F": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`EC2 .createProdVpc() matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PreFixVpc359F0D17": {
      "Properties": {
        "CidrBlock": "10.255.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "PreFixVpcIGWB72EC1F8": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "PreFixVpcPublicSubnet1DefaultRouteB716A66A": {
      "DependsOn": [
        "PreFixVpcVPCGWC09F8D3F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet1RouteTable23041149",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PreFixVpcPublicSubnet1RouteTable23041149": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PreFixVpcPublicSubnet1RouteTableAssociation0CC9D25D": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet1RouteTable23041149",
        },
        "SubnetId": {
          "Ref": "PreFixVpcPublicSubnet1Subnet8F335C33",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PreFixVpcPublicSubnet1Subnet8F335C33": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.255.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PreFixVpcPublicSubnet2DefaultRoute41C7E8D0": {
      "DependsOn": [
        "PreFixVpcVPCGWC09F8D3F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet2RouteTableF7146078",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PreFixVpcPublicSubnet2RouteTableAssociationB661D737": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet2RouteTableF7146078",
        },
        "SubnetId": {
          "Ref": "PreFixVpcPublicSubnet2Subnet387DF9A4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PreFixVpcPublicSubnet2RouteTableF7146078": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PreFixVpcPublicSubnet2Subnet387DF9A4": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.255.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PreFixVpcVPCGWC09F8D3F": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`EC2 .createVpc() matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "PreFixVpc359F0D17": {
      "Properties": {
        "CidrBlock": "10.255.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "PreFixVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "PreFixVpcIGWB72EC1F8": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "PreFixVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "PreFixVpcPublicSubnet1DefaultRouteB716A66A": {
      "DependsOn": [
        "PreFixVpcVPCGWC09F8D3F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet1RouteTable23041149",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PreFixVpcPublicSubnet1RouteTable23041149": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PreFixVpcPublicSubnet1RouteTableAssociation0CC9D25D": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet1RouteTable23041149",
        },
        "SubnetId": {
          "Ref": "PreFixVpcPublicSubnet1Subnet8F335C33",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PreFixVpcPublicSubnet1Subnet8F335C33": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.255.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PreFixVpcPublicSubnet2DefaultRoute41C7E8D0": {
      "DependsOn": [
        "PreFixVpcVPCGWC09F8D3F",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet2RouteTableF7146078",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "PreFixVpcPublicSubnet2RouteTableAssociationB661D737": {
      "Properties": {
        "RouteTableId": {
          "Ref": "PreFixVpcPublicSubnet2RouteTableF7146078",
        },
        "SubnetId": {
          "Ref": "PreFixVpcPublicSubnet2Subnet387DF9A4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "PreFixVpcPublicSubnet2RouteTableF7146078": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "PreFixVpcPublicSubnet2Subnet387DF9A4": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.255.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "Default/PreFixVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "PreFixVpcVPCGWC09F8D3F": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "PreFixVpcIGWB72EC1F8",
        },
        "VpcId": {
          "Ref": "PreFixVpc359F0D17",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
